crates/compiler/constrain/src/expr.rs

```diff
@@ -375,6 +375,38 @@ pub fn constrain_expr(
             let expected_index = expected;
             constraints.equal_types(str_index, expected_index, Category::Str, region)
         }
+        IngestedFile(bytes, anno) => match &anno.typ {
+            Type::Apply(Symbol::STR_STR, _, _) => {
+                if std::str::from_utf8(bytes).is_err() {
+                    todo!("cause an error for the type being wrong due to not being a utf8 string");
```

Comment:
@ayazhafiz, How would suggest I make this bubble up an error? Or I guess, where else could I put this check that can bubble up an error?

Reply:
I think this check should happen in the solver, and not the constraint generation. Instead of conditionally determining the constraint to create here, what do you think about creating a new Constraint variant called `IngestedFile`? Then, in solve/solve, you implement the variant by trying to unify against a `Str`; if that fails, unify against a `List U8` right there. If both fail, you can then push a `TypeProblem` that represents "this ingested file was not a Str or List U8".

That will make sure we deal with aliases, etc. correctly as well.

Reply:
Ok. That does make things way nicer. I have changed to this and I think the PR is much closer to ready.

That said, part of may change broke alias analysis. So need to dig into that.

Reply:
Hmm. The alias analysis error is due to calling unify multiple types in a row. I guess the first call to unify modifies the subs/env. I guess I either need a version of unify that doesn't modify the subs or need to clone the subs each time. Maybe there is some other alternative.

Reply:
`subs.snapshot/rollback` if the unification fails is reasonable.

Reply:
Thanks so much. Works great.